dev/src/path.ts

```diff
@@ -199,89 +185,132 @@ abstract class Path<T> {
 }
 
 /**
- * A slash-separated path for navigating resources (documents and collections)
- * within Firestore.
+ * A slash-separated path for navigating resources within the current Firestore
+ * instance.
  *
  * @private
- * @class
  */
-export class ResourcePath extends Path<ResourcePath> {
+export class RelativePath extends Path<RelativePath> {
```

Comment:
So I'm having some trouble with this RelativePath class:
1. It's not an _arbitrary_ relative path (else isDocument() and isCollection() wouldn't be answerable).
2. "Relative" probably isn't the defining characteristic of this class.  In fact, the `Path` base class is already a relative path.  It seems like the unique methods (`id()`, `isCollection()`, `isDocument()`) are really characteristics of being a "resource path" rather than a "relative path."
3. It seems strange for ResourcePath to extend RelativePath.  "A resource path **is a** relative path" doesn't hold. And it's confusing for a method that takes `path: RelativePath` that `path` may actually be an (absolute) `ResourcePath`.

I haven't analyzed the code to figure out how these classes are used and what other options might work. Naive idea fragments:
1. I think the mobile SDKs deal with this by having all ResourcePaths be relative and we only create an "absolute" one at the point we serialize.  But that has downsides (most notably we can't handle resource-paths in documents that point to to other projects).
2. If there's a way to solve this via composition rather than inheritance (e.g. have a `DatabasePath` that contains projectId, databaseId, and resourcePath members), I'd be a fan. I haven't read it, but presumably this wikipedia article makes the argument for me about why composition should be preferred over inheritance: https://en.wikipedia.org/wiki/Composition_over_inheritance
3. Failing all that, if you renamed ResourcePath to AbsoluteResourcePath and RelativePath to ResourcePath that would actually solve most of the problems I listed above I think.

Perhaps any of these solutions will involve a lot of churn though, so we may want to punt to a separate PR.

Reply:
I kept the class structure as is. I actually think that inheritance makes sense here - we are special casing the RelativePath class and modify the existing behavior only slightly. Furthermore, we can use both classes interchangeably, without having to to rely on TypeScrips's intersection types (or introduce another interface).

With all that being said, I renamed the classes to ResourcePath and QualifiedResourcePath. This should address most of your concerns. While there is some churn, it's 70ish lines and I pushed in a single commit together with the rest of the feedback (ae028e548ea87a734c677808736b9c1b0ce1b9e1).